src/common/libsdexec/test/channel_outbuf.c

```diff
@@ -0,0 +1,217 @@
+    int len;
+    bool eof;
+
+    if (iodecode (io, &stream, NULL, NULL, &len, &eof) < 0) {
+        diag ("%s: idoecode error: %s",
+              sdexec_channel_get_name (ch),
+              strerror (errno));
+    }
+    else {
+        diag ("%s output: stream=%s len=%d eof=%s",
+              sdexec_channel_get_name (ch),
+              stream,
+              len,
+              eof ? "true" : "false");
+        raw_byte_count += len;
+        if (eof)
+            flux_reactor_stop (r);
+    }
+}
+
+void test_raw (size_t bufsize, size_t datasize)
+{
+    flux_t *h;
+    struct channel *ch;
+    int fd;
+    char buf[datasize];
+    flux_reactor_t *r;
+
+    if (!(h = flux_open ("loop://", 0)))
+        BAIL_OUT ("could not create loop flux_t handle for testing");
+    r = flux_get_reactor (h);
+    if (flux_attr_set_cacheonly (h, "rank", "0") < 0)
+        BAIL_OUT ("could not set rank for testing");
+    ch = sdexec_channel_create_output (h,
+                                       "raw",
+                                       bufsize,
+                                       0,
+                                       raw_output_cb,
+                                       error_cb,
+                                       r);
+    ok (ch != NULL,
+        "sdexec_channel_crate_output works");
+    sdexec_channel_start_output (ch);
+    ok (true, "sdexec_channel_start_output called");
+
+    fd = sdexec_channel_get_fd (ch);
+    ok (fd >= 0,
+        "sdexec_channel_get_fd works");
+
+    raw_byte_count = 0;
+    error_called = false;
+    // write can exceed TINY_OUTBUF size by approx O(PAGE_SIZE) socket buf
+    memset (buf, 'x', sizeof (buf));
+    ok (write (fd, buf, sizeof (buf)) == sizeof (buf),
+        "wrote %zu bytes of data from unit", sizeof (buf));
+    sdexec_channel_close_fd (ch);
+    ok (true, "sdexec_channel_close_fd called");
+    ok (flux_reactor_run (r, 0) >= 0,
+        "flux_reactor_run ran successfully");
+    ok (error_called == false,
+        "error callback was not called");
+    ok (raw_byte_count == sizeof (buf),
+        "all bytes were received");
+
+    sdexec_channel_destroy (ch);
+    flux_close (h);
+}
+
+size_t line_byte_count;
+size_t line_count;
+
+void line_output_cb (struct channel *ch, json_t *io, void *arg)
+{
+    flux_reactor_t *r = arg;
+    const char *stream;
+    char *data = NULL;
+    int len;
+    bool eof;
+
+    if (iodecode (io, &stream, NULL, &data, &len, &eof) < 0) {
+        diag ("%s: idoecode error: %s",
+              sdexec_channel_get_name (ch),
+              strerror (errno));
+    }
+    else {
+        diag ("%s output: stream=%s len=%d eof=%s",
+              sdexec_channel_get_name (ch),
+              stream,
+              len,
+              eof ? "true" : "false");
+        line_byte_count += len;
+        int count = 0;
+        for (int i = 0; i < len; i++)
+            if (data[i] == '\n')
+                count++;
+        line_count += count;
```

Comment:
shouldn't it be at most one line per callback?  I would think > 1 line should be an error?

Reply:
Yes.  Whoops - I messed that up.  Good catch!

Reply:
Just pushed fixes for those comments.

> shouldn't it be at most one line per callback? I would think > 1 line should be an error?

Sort of, usually.  There will be one extra callback for eof.  And if the line length exceeds the buffer size, the callback will get buffer sized chunks without newlines until the newline is received.  I noted that in the test.

Reply:
ahhh ok, I guess when I look at

```
        for (int i = 0; i < len; i++) {
            if (data[i] == '\n')
                count++;
        }
        line_count += count;
```

it suggests to me that number of lines could be > 1 in a single callback, which I think is impossible?  I would think it should be more like

```
        for (int i = 0; i < len; i++) {
            if (data[i] == '\n')
                count++;
        }
        if (count > 1) {
            /* some error */
        }
```

Reply:
The test is structured to keep separate tallies of lines and callbacks, then it checks them against the expected number at the end.  So I think that case is covered - if two lines were returned in one callback, the line count would be correct but the callback count would be wrong.

I guess my thinking was that throwing one error at the end made more sense than calling out the same error possibly multiple times.